### ISSUE TITLE
Loading time can take up to one hour

### ISSUE THREAD

**Issue by thinkingserious** [stoplightio/prism#1743]
## Describe the bug

There is a bug in the v8 engine that was to be resolved in the latest version. After upgrading, the bug persists.

Please see: https://github.com/stoplightio/prism/issues/1479#issue-722791366 for further details

## To Reproduce

Please see: https://github.com/stoplightio/prism/issues/1479

## Expected behavior

Please see: https://github.com/stoplightio/prism/issues/1479#issue-722791366

## Additional context
Creating a new issue based on the last interaction at #1479

**Comment by marbemac** [comment]
Long time no talk @thinkingserious!

@karol-maciaszek looking at https://github.com/stoplightio/prism/issues/1479#issuecomment-735983921 - why are we including / using query parameters when matching route in the first place? Usually the only bit relevant to routing is the path.

**Comment by marbemac** [comment]
And another q re routing @karol-maciaszek - had we ever considered using a battle hardened lib like https://github.com/pillarjs/path-to-regexp rather than rolling our own?

**Comment by X3per** [comment]
> And another q re routing @karol-maciaszek - had we ever considered using a battle hardened lib like https://github.com/pillarjs/path-to-regexp rather than rolling our own?



> And another q re routing @karol-maciaszek - had we ever considered using a battle hardened lib like https://github.com/pillarjs/path-to-regexp rather than rolling our own?

`uri-template-lite` is being used for this regex generation. `path-to-regex` might help here but only if it produces different regex for the same input. The main problem may still reside inside node@15 engine.

I'd try two things:
- re-do the reproduction code provided by. @XVincentX here: https://github.com/stoplightio/prism/issues/1479#issuecomment-735983921
- try to replace `uri-template-lite` with `path-to-regex` (don't like it, we do use `uri-template-lite` in other places, we'll end up with two libs doing the same)

**Comment by karol-maciaszek** [comment]
> And another q re routing @karol-maciaszek - had we ever considered using a battle hardened lib like https://github.com/pillarjs/path-to-regexp rather than rolling our own?



> And another q re routing @karol-maciaszek - had we ever considered using a battle hardened lib like https://github.com/pillarjs/path-to-regexp rather than rolling our own?

`uri-template-lite` is being used for this regex generation. `path-to-regex` might help here but only if it produces different regex for the same input. The main problem may still reside inside node@15 engine.

I'd try two things:
- re-do the reproduction code provided by. @XVincentX here: https://github.com/stoplightio/prism/issues/1479#issuecomment-735983921
- try to replace `uri-template-lite` with `path-to-regex` (don't like it, we do use `uri-template-lite` in other places, we'll end up with two libs doing the same)

**Comment by XVincentX** [comment]
Folks! 👋 

Since I got summoned here — I took some time to dig into and yes, the backtracking regex improvements originally scheduled for Node 15 have been delayed for now (as in, Node 15 does not include the required V8 version).

Even though I do not really have a say in the matter anymore — I agree with @karol-maciaszek and would strongly suggest to hold off on changing anything in the code.

This is a platform problem that's really close to be fixed. I also suggested a temporary workaround if required.

You people decide what to do.

**Comment by marbemac** [comment]
Ok, any extra thought re question in https://github.com/stoplightio/prism/issues/1743#issuecomment-816107624?

**Comment by thinkingserious** [comment]
Thank you for the additional feedback @karol-maciaszek & @XVincentX!

Hi @marbemac!

Are you referring to 
> "why are we including / using query parameters when matching route in the first place? Usually the only bit relevant to routing is the path."

**Comment by marbemac** [comment]
Yeah - at first glance it does not seem necessary, but wanted to see if @karol-maciaszek or @XVincentX had any extra context re why that's necessary.

**Comment by XVincentX** [comment]
> why are we including / using query parameters when matching route in the first place? 

@marbemac 

We aren't. Prism Http does *not* have this problem, this only happens when using the CLI, and it happens when trying to generate examples to show in the CLI output. https://github.com/stoplightio/prism/blob/ae5dcdf5ebbe590fa095fa9a49c0378295928e5e/packages/cli/src/util/paths.ts#L31

**Comment by radzserg** [comment]
I tried to test the problematic regular expression with the latest node version 16.0.0. The problem did not disappear, but only got worse compared to version 10.*.*
```
➜  node_test cat index.js
const template = '/2010-04-01/Accounts/XXXXXX/Calls.json{?ParentCallSid,StartTime,StartTime<,StartTime>}';
const re = /{([#&+.\/;?]?)((?:[-\w%.]+(\*|:\d+)?,?)+)}/g
console.time();
console.log("match", template.match(re))
console.timeEnd();

➜  node_test node index.js
match null
default: 39.378s

➜  node_test node -v
v14.16.1
➜  node_test node index.js
match null
default: 37.570s

➜  node_test node -v
v10.13.0
➜  node_test node index.js
match null
default: 7006.431ms
```

This makes me think we need to find an alternative that doesn't use regular expressions. Stay tuned... I'll provide more feedback soon.

**Comment by marbemac** [comment]
@radzserg @lottamus do I understand right that this problematic functionality is only in place so that we can include the query string when we print out example urls to the cli console? IF that is the case, then I think it's a worthwhile trade to adjust what is printed to the console to solve the issue (by not including query string in printed example urls, assuming that's what this relates to).

**Comment by radzserg** [comment]
I actually found out that the cause is special characters `>` & `<` that we have in param names. And strictly speaking, they shouldn't be allowed. Compare 2 results when params are raw and encoded 

```
const template = '/2010-04-01/Accounts/XXXXXX/Calls.json{?ParentCallSid,StartTime,StartTime<,StartTime>}';
const transformedValues = {
    "ParentCallSid": "~pre~CAb1d0eAEC7122DeBA19eab3EE61d6D88c~post~",
    "StartTime": "~pre~1971-02-07T22:53:40.662Z~post~",
    "StartTime<": "~pre~1978-02-03T22:36:00.877Z~post~",
    "StartTime>": "~pre~1985-10-25T03:33:00.613Z~post~"
}
console.time('raw');
console.log(URI.expand(template, transformedValues));
console.timeEnd('raw');


const templateEncoded = '/2010-04-01/Accounts/XXXXXX/Calls.json{?ParentCallSid,StartTime,StartTime%3C,StartTime%3E}';
const transformedValuesEncoded = {
    "ParentCallSid": "~pre~CAb1d0eAEC7122DeBA19eab3EE61d6D88c~post~",
    "StartTime": "~pre~1971-02-07T22:53:40.662Z~post~",
    "StartTime%3C": "~pre~1978-02-03T22:36:00.877Z~post~",
    "StartTime%3E": "~pre~1985-10-25T03:33:00.613Z~post~"
}
console.time('encoded');
console.log(URI.expand(templateEncoded, transformedValuesEncoded));
console.timeEnd('encoded');

```
**Results**
```
/2010-04-01/Accounts/XXXXXX/Calls.json{?ParentCallSid,StartTime,StartTime<,StartTime>}
raw: 7.389s
/2010-04-01/Accounts/XXXXXX/Calls.json?ParentCallSid=~pre~CAb1d0eAEC7122DeBA19eab3EE61d6D88c~post~&StartTime=~pre~1971-02-07T22%3A53%3A40.662Z~post~&StartTime%3C=~pre~1978-02-03T22%3A36%3A00.877Z~post~&StartTime%3E=~pre~1985-10-25T03%3A33%3A00.613Z~post~
encoded: 0.459ms
```

Raw params expanded incorrectly.  I tried to test expand with other libraries and all of them complained about `>` `<` characters. Also as you can see time spent on `expand` with encoding params is significantly reduced. I'm going to make a fix with encoding so we can save functionality.

**Comment by radzserg** [comment]
@thinkingserious I would be grateful if you can test [my changes](https://github.com/stoplightio/prism/pull/1816) in the original file. I don't have access to it, but I'm curious how the changes would work on a large file.
You can test it by
```
git checkout fix/1743-loading-time
cd packages/cli 
yarn cli mock /path/to/spec
```

**Comment by thinkingserious** [comment]
For reference, this is the updated file: https://github.com/twilio/twilio-oai/blob/main/spec/json/twilio_api_v2010.json

Thank you!

**Comment by thinkingserious** [comment]
Hello @radzserg,

After executing the steps above, I get the following error:

```
git checkout fix/1743-loading-time
cd packages/cli 
yarn cli mock /path/to/spec
```

```
yarn run v1.22.10
$ node -r ts-node/register/transpile-only -r tsconfig-paths/register src/index.ts mock /Users/ethomas/di/github/twilio-oai/spec/json/twilio_api_v2010.json
node:internal/modules/cjs/loader:943
  throw err;
  ^

Error: Cannot find module 'ts-node/register/transpile-only'
Require stack:
- internal/preload
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:940:15)
    at Function.Module._load (node:internal/modules/cjs/loader:773:27)
    at Module.require (node:internal/modules/cjs/loader:1012:19)
    at Module._preloadModules (node:internal/modules/cjs/loader:1260:12)
    at loadPreloadModules (node:internal/bootstrap/pre_execution:456:5)
    at prepareMainThreadExecution (node:internal/bootstrap/pre_execution:75:3)
    at node:internal/main/run_main_module:7:1 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ 'internal/preload' ]
}
error Command failed with exit code 1.
```

**Comment by thinkingserious** [comment]
I did not have yarn installed 🤦🏼 , after installing yarn and running `yarn cli mock /path/to/spec` the spec loaded almost instantly 🎉

**Comment by radzserg** [comment]
Cool, I've released the latest changes in v4.2.2
